src/tools/cgclassify.c

```diff
@@ -198,7 +231,459 @@ int main(int argc, char *argv[])
+		 * /proc/mounts
+		 */
+		find_mnt_point("unified", &cgroup_name);
+		if (!cgroup_name) {
+			err("Unable find unified cgroup path\n");
+			fclose(cgroup_systemd_path_f);
+			return -1;
+		}
+
+		snprintf(cgroup_procs_path, sizeof(cgroup_procs_path), "%s/%s/cgroup.procs",
+			 cgroup_name, scope_name);
+		free(cgroup_name);
+
+		cgroup_unified_path_f = fopen(cgroup_procs_path, "we");
+		if (!cgroup_unified_path_f) {
+			err("Failed to open %s\n", cgroup_procs_path);
+			fclose(cgroup_systemd_path_f);
+			return -1;
+		}
+	}
+
+	fprintf(cgroup_systemd_path_f, "%d", pid);
+	fflush(cgroup_systemd_path_f);
+	fclose(cgroup_systemd_path_f);
+
+	if (!is_cgroup_mode_hybrid())
+		return 0;
+
+	fprintf(cgroup_unified_path_f, "%d", pid);
+	fflush(cgroup_unified_path_f);
+	fclose(cgroup_unified_path_f);
+
+	return 0;
+}
+
+static int is_scope_parsed(const char * const path)
+{
+	/*
+	 * As per, <kernel sources>/kernel/cgroup/cgroup.c::cgroup_init()
+	 * At the max there can be only 16 controllers and we are
+	 * not accounting for named hierarchies, which can be more
+	 * than 16 themselves.
+	 */
+	static char parsed_scope_path[MAX_MNT_ELEMENTS][FILENAME_MAX];
+	int i;
+
+	for (i = 0; i < MAX_MNT_ELEMENTS; i++) {
+		if (!strcmp(parsed_scope_path[i], path))
+			return 1;
+		if (parsed_scope_path[i][0] == '\0') {
+			snprintf(parsed_scope_path[i], FILENAME_MAX, "%s", path);
+			break;
+		}
+	}
+
+	return 0;
+}
+
+/* Borrowed from src/api.c::__attach_task_pid */
+static int attach_task_pid(char *path, pid_t tid)
+{
+	FILE *tasks = NULL;
+	int ret = 0;
+
+	tasks = fopen(path, "we");
+	if (!tasks) {
+		switch (errno) {
+		case EPERM:
+			ret = ECGROUPNOTOWNER;
+			break;
+		case ENOENT:
+			ret = ECGROUPNOTEXIST;
+			break;
+		default:
+			ret = ECGROUPNOTALLOWED;
+		}
+		goto err;
+	}
+	ret = fprintf(tasks, "%d", tid);
+	if (ret < 0) {
+		ret = ECGOTHER;
+		goto err;
+	}
+	ret = fflush(tasks);
+	if (ret) {
+		ret = ECGOTHER;
+		goto err;
+	}
+	fclose(tasks);
+	return 0;
+err:
+	err("cannot write tid %d to %s:%s\n", tid, path, strerror(errno));
+	if (tasks)
+		fclose(tasks);
+	return ret;
+}
+
+static int rollback_pid_cgroups(pid_t pid)
+{
+	char cgroup_proc_path[FILENAME_MAX + 14];
+	char cgroup_path[FILENAME_MAX];
+	int err = 0, idx = 0, ret = 0;
+	char *cgrp_proc_path = NULL;
+
+	/*
+	 * unified cgroup rollback is simple, we need to write into
+	 * single cgroup hierarchy.
+	 */
+	if (is_cgroup_mode_unified()) {
+		pthread_rwlock_rdlock(&cg_mount_table_lock);
+		cg_build_path_locked(info[idx].cgrp_path, cgroup_path, NULL);
+		pthread_rwlock_unlock(&cg_mount_table_lock);
+
+		snprintf(cgroup_proc_path, FILENAME_MAX + 14, "%s/cgroup.procs", cgroup_path);
+		ret = attach_task_pid(cgroup_proc_path, pid);
+		return ret;
+	}
+
+	for (idx = 0; info[idx].ctrl_name[0] != '\0'; idx++) {
+		/* find the systemd cgroup path */
+		if (!strcmp(info[idx].ctrl_name, "name=systemd")) {
+			find_mnt_point("name=systemd ", &cgrp_proc_path);
```

Comment:
You have a space at the end of `name=systemd `.  If a system ever places the `name=systemd` option before `noexec`, `nodev`, etc. then it won't be found.  Is this by design?

Reply:
@drakenclimber Thank you so much for the review. Agreed, it's not robust will come up with something better.